src/haddock/modules/refinement/cgtoaa/__init__.py

```diff
@@ -0,0 +1,126 @@
+implemented in CNS.
+"""
+
+from pathlib import Path
+
+from haddock.core.defaults import MODULE_DEFAULT_YAML
+from haddock.core.typing import FilePath
+from haddock.gear.haddockmodel import HaddockModel
+from haddock.libs.libcns import prepare_cns_input, prepare_expected_pdb
+from haddock.libs.libontology import PDBFile
+from haddock.libs.libsubprocess import CNSJob
+from haddock.modules import get_engine
+from haddock.modules.base_cns_module import BaseCNSModule
+
+
+RECIPE_PATH = Path(__file__).resolve().parent
+DEFAULT_CONFIG = Path(RECIPE_PATH, MODULE_DEFAULT_YAML)
+
+
+class HaddockModule(BaseCNSModule):
+    """HADDOCK3 module energy minimization refinement."""
+
+    name = RECIPE_PATH.name
+
+    def __init__(
+        self, order: int, path: Path, initial_params: FilePath = DEFAULT_CONFIG
+    ) -> None:
+        """."""
+        cns_script = Path(RECIPE_PATH, "cns", "cgtoaa.cns")
+        super().__init__(order, path, initial_params, cns_script=cns_script)
+
+    @classmethod
+    def confirm_installation(cls) -> None:
+        """Confirm module is installed."""
+        return
+
+    def _run(self) -> None:
+        """Execute module."""
+        # Pool of jobs to be executed by the CNS engine
+        jobs: list[CNSJob] = []
+
+        # Get the models generated in previous step
+        try:
+            models_to_refine = self.previous_io.retrieve_models()
+        except Exception as e:
+            self.finish_with_error(e)
+
+        self.output_models = []
+        sampling_factor = self.params["sampling_factor"]
+        if sampling_factor > 1:
+            self.log(f"sampling_factor={sampling_factor}")
+        if sampling_factor == 0:
+            self.log("[Warning] sampling_factor cannot be 0, setting it to 1")
+            sampling_factor = 1
+        if sampling_factor > 100:
+            self.log("[Warning] sampling_factor is larger than 100")
+
+        max_nmodels = self.params["max_nmodels"]
+        nmodels = len(models_to_refine) * sampling_factor
+        if nmodels > max_nmodels:
+            self.finish_with_error(
+                f"Too many models ({nmodels}) to refine, max_nmodels ="
+                f" {max_nmodels}. Please reduce the number of models or"
+                " decrease the sampling_factor."
+            )
+
+        model_idx = 0
+        idx = 1
+        for model in models_to_refine:
+            model_idx += 1
+
+            #print(self.params["sampling_factor"])
+            for _ in range(self.params["sampling_factor"]):
+                cgtoaa_input = prepare_cns_input(
+                    idx,
+                    model,
+                    self.path,
+                    self.recipe_str,
+                    self.params,
+                    "cgtoaa",
+                    native_segid=True,
+                    debug=self.params["debug"],
+                    seed=model.seed if isinstance(model, PDBFile) else None,
+                    cgtoaa=True
+                )
```

Comment:
I think this seed value must be modified for each sampled models `sampling_factor`
```suggestion
            for s_ind in range(self.params["sampling_factor"]):
                cgtoaa_input = prepare_cns_input(
                    idx,
                    model,
                    self.path,
                    self.recipe_str,
                    self.params,
                    "cgtoaa",
                    native_segid=True,
                    debug=self.params["debug"],
                    seed=(model.seed + s_ind) if isinstance(model, PDBFile) else None,
                    cgtoaa=True
                )
```

Reply:
This was copied from the emref module, should it also be changed in the emref then ?

Reply:
In minimisation there is no seed used. For the cgtoaa module we use molecular dynamics and thus a starting seed. So only needed here.  You should also find it in the mdref module for example